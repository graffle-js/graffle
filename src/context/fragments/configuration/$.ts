export * as Configuration from './$$.js'
