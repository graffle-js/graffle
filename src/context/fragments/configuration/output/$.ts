export * as Output from './$$.js'
