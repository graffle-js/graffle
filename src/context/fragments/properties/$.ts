export * as Properties from './$$.js'
