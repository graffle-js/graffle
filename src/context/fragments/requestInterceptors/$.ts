export * as RequestInterceptors from './$$.js'
