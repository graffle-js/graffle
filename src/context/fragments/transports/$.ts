export * as Transports from './$$.js'
