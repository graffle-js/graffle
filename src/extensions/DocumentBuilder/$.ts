/**
 * @deprecated DocumentBuilderKit is deprecated. Use Docpar.Object instead.
 *
 * ```diff
 * - import { DocumentBuilderKit } from 'graffle'
 * + import { Docpar } from 'graffle'
 * ```
 *
 * Mapping:
 * - DocumentBuilderKit.Select → Docpar.Object.Select
 * - DocumentBuilderKit.Var → Docpar.Object.Var
 * - DocumentBuilderKit.InferResult → Docpar.Object.InferResult
 * - DocumentBuilderKit.ToGraphQLDocument → Docpar.Object.ToGraphQLDocument
 */
export { Object as DocumentBuilderKit } from '#src/docpar/$.js'
