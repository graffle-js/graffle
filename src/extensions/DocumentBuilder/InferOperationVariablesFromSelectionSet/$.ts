export * as InferOperationVariablesFromSelectionSet from './$$.js'
