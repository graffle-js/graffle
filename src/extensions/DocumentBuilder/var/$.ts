export * as Var from './$$.js'
