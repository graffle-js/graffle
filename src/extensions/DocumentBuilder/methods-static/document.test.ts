import type { TypedFullDocument } from '#src/lib/grafaid/typed-full-document/$.js'
import { Ts } from '@wollybeard/kit'
import { Test } from '@wollybeard/kit/test'
import { expect, test } from 'vitest'
import { Possible } from '../__tests__/fixtures/possible/$.js'
import { Var } from '../var/$.js'
import type { InferOperations } from './document.js'

const $ = Var.$
type $ = typeof $

// dprint-ignore
Test.describe('static document builder')
  .i <TypedFullDocument.TypedFullDocumentString>()
  .o<{ y: string[]; n?: string[] }>()
  .cases(
    [[Possible.document({ query: { getUser: { id: true, idNonNull: true } } })],                                                   { y: [`query getUser`, `id`, `idNonNull`] }],
    [[Possible.document({ query: { getUser: { id: true }, getPost: { idNonNull: true } } })],                                      { y: [`query getUser`, `query getPost`, `id`, `idNonNull`] }],
    [[Possible.document({ query: { getUserWithArg: { stringWithRequiredArg: { $: { string: $ } } } } })],                          { y: [`query getUserWithArg`, `$string:`, `stringWithRequiredArg`] }],
    [[Possible.document({ query: { getUser: { id: true } }, mutation: { updateUser: { id: true } } })],                            { y: [`query getUser`, `mutation updateUser`, `id`] }],
  )
  .test((doc, expectations) => {
    for (const y of expectations.y) {
      expect(doc).toContain(y)
    }

    if (expectations.n) {
      for (const n of expectations.n) {
        expect(doc).not.toContain(n)
      }
    }
  })

test('document is type checked', () => {
  Possible.document({
    // @ts-expect-error invalid document - $ syntax requires Var.$ marker, not boolean literal
    query: { foo: { stringWithRequiredArg: { $: { string: true /* <-- should be $ */ } } } },
  })
})

// ====================================================================
//                   INFER OPERATIONS TYPE TESTS
// ====================================================================
// dprint-ignore
type _1 = Ts.Cases<
  // InferOperations with both query and mutation operations
  Ts.AssertEqual<
    InferOperations<
      {
        query: {
          getUser: { id: true }
        }
        mutation: {
          updateUser: { idNonNull: true }
        }
      },
      Possible.$.Schema,
      Possible.$.ArgumentsMap,
      { typeHookRequestResultDataTypes: never; scalars: {} }
    >,
    {
      getUser: { result: { id: string | null }; variables: {} }
      updateUser: { result: { idNonNull: string }; variables: {} }
    }
  >,
  // InferOperations with mutation variables including $-prefixed keys
  Ts.AssertEqual<
    InferOperations<
      {
        mutation: {
          addItem: {
            id: {
              $: {
                $type: $
                name: $
              }
            }
          }
        }
      },
      Possible.$.Schema,
      Possible.$.ArgumentsMap,
      { typeHookRequestResultDataTypes: never; scalars: {} }
    >,
    {
      addItem: {
        result: { id: string | null }
        variables: { type?: string | undefined; name?: string | undefined }
      }
    }
  >
>
