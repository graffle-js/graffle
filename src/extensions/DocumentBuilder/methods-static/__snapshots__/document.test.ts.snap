// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`static document builder > document({ query: { getUser: { id: true } }, mutation: { updateUser: { id: true } } }) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
{
  query: {
    getUser: {
      id: true
    }
  },
  mutation: {
    updateUser: {
      id: true
    }
  }
}
╠══════════════════════════════════════════════════╣ THEN RETURNS STRING
query getUser {
  id
}

mutation updateUser {
  id
}
╚══════════════════════════════════════════════════╝"
`;

exports[`static document builder > document({ query: { getUser: { id: true }, getPost: { idNonNull: true } } }) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
{
  query: {
    getUser: {
      id: true
    },
    getPost: {
      idNonNull: true
    }
  }
}
╠══════════════════════════════════════════════════╣ THEN RETURNS STRING
query getUser {
  id
}

query getPost {
  idNonNull
}
╚══════════════════════════════════════════════════╝"
`;

exports[`static document builder > document({ query: { getUser: { id: true, idNonNull: true } } }) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
{
  query: {
    getUser: {
      id: true,
      idNonNull: true
    }
  }
}
╠══════════════════════════════════════════════════╣ THEN RETURNS STRING
query getUser {
  id
  idNonNull
}
╚══════════════════════════════════════════════════╝"
`;

exports[`static document builder > document({ query: { getUserWithArg: { stringWithRequiredArg: { $: { string: [Function: na...) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
{
  query: {
    getUserWithArg: {
      stringWithRequiredArg: {
        $: {
          string: [Function: name(name) {
      return self({
        ...state,
        name
      });
    }] { name: [Function: name], _: [Object], as: [Function: as], default: [Function: default], required: [Function: required] }
        }
      }
    }
  }
}
╠══════════════════════════════════════════════════╣ THEN RETURNS STRING
query getUserWithArg($string: String!) {
  stringWithRequiredArg(string: $string)
}
╚══════════════════════════════════════════════════╝"
`;
